cardano-node/src/Cardano/Node/Run.hs

```diff
@@ -326,20 +357,49 @@ handleSimpleNode p trace nodeTracers npm onKernel = do
 
          when mockValidateDB $ traceWith tracer "Performing DB validation"
 
+  stopOnSlotSynced :: NodeProtocolMode -> MaxSlotNo
+  stopOnSlotSynced (RealProtocolMode cli) = shutdownOnSlotSynced cli
+  stopOnSlotSynced _ = NoMaxSlotNo
+
+  -- | Produce a potential pair of FDs:
+  --  - read end, where EOF is a signal of termination
+  --  - write end, which is means of internal termination handling
+  decideShutdownFds :: NodeProtocolMode -> IO ShutdownFDs
+  decideShutdownFds (RealProtocolMode NodeCLI{shutdownIPC = Just fd}) = pure $ ExternalShutdown fd
+  decideShutdownFds (RealProtocolMode NodeCLI{shutdownOnSlotSynced = MaxSlotNo{}}) = do
+    (r, w) <- IO.createPipeFd
+    pure $ InternalShutdown (Fd r) (Fd w)
+  decideShutdownFds _ = pure NoShutdownFDs
+
+data ShutdownFDs
+  = NoShutdownFDs
+  | ExternalShutdown Fd
+  -- ^ We shall watch this externally-supplied FD for termination EOF.
+  | InternalShutdown Fd Fd
+  -- ^ Pair of FDs for intra-process graceful shutdown termination messaging.
+  --   First FD is the read side as above,
+  --   second FD is the shutdown doorbell to be closed.
+
+sfdsReadSide :: ShutdownFDs -> Maybe Fd
+sfdsReadSide (ExternalShutdown r) = Just r
+sfdsReadSide (InternalShutdown r _w) = Just r
+sfdsReadSide _ = Nothing
```

Comment:
I recently saw a PR about cleanly shutting down the node go by, but I'm not familiar with the details. Can somebody more knowledgeable about it review this part?

Reply:
@mrBliss, that was @dcoutts 's work, and I'm reusing & extending his machinery a bit.